course6/toyapp-adam/util/keyboard.h

```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include <cstdint>
+#include <SDL2/SDL.h>
+
+namespace Util {
+
+class Keyboard
+{
+public:
+	using KeyCode = SDL_Keycode;
+
+	void Press(KeyCode keyCode);
+	void Release(KeyCode keyCode);
+	void ClearChanges();
+	bool IsUp(KeyCode keyCode) const;
+	bool IsDown(KeyCode keyCode) const;
+	bool IsPressed(KeyCode keyCode) const;
```

Comment:
What's the difference between ```Down``` and ```Pressed```.
This is API is a gross violation of the principle of the least surprise.

Reply:
Difference is that a key might be in a pressed (down) state, but then we can't know whether it has just been pressed in the current Update.

Since our project is a simulation running in a loop, in my opinion it's perfectly valid for the codebase to be aware that something could be related to the current Update only (pressed, released), or a state of the simulation in general (down, up).

Indeed, input could be event-based only instead. That's a design decision and for now I chose polling key states instead. I think having _both_ events and checking state would be the optimal solution.